In [ ]:
#pip install nltk

Note: you may need to restart the kernel to use updated packages.


e:\Gen AI Bootcamp\Full-Stack-GenerativeAI-And-Agentic-AI-Bootcamp\myenv\Scripts\python.exe: No module named pip


In [3]:
import pandas as pd
import numpy as np
import os
import gensim
import nltk

In [4]:
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     C:\Users\ICON\AppData\Roaming\nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     C:\Users\ICON\AppData\Roaming\nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     C:\Users\ICON\AppData\Roaming\nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     C:\Users\ICON\AppData\Roaming\nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_eng is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     C:\Users\ICON\AppData\Roaming\nltk_data...
[nltk_

True

In [27]:
from nltk import sent_tokenize
from gensim.utils import simple_preprocess

In [28]:
for file in os.listdir('./data/'):
    print(file)

story1.txt
story2.txt


In [29]:
f = open(r"E:\Gen AI Bootcamp\Full-Stack-GenerativeAI-And-Agentic-AI-Bootcamp\Class 06_11April26 word2vec, avg word2vec practical\Data\story1.txt",encoding='unicode_escape')
f

<_io.TextIOWrapper name='E:\\Gen AI Bootcamp\\Full-Stack-GenerativeAI-And-Agentic-AI-Bootcamp\\Class 06_11April26 word2vec, avg word2vec practical\\Data\\story1.txt' mode='r' encoding='unicode_escape'>

In [30]:
corpus = f.read()
corpus

'Once upon a time, there was a young boy named Arjun who lived in a small town surrounded by hills and rivers. He was very curious about how things worked and always asked questions. His parents often smiled at his curiosity and encouraged him to learn more.\n\nEvery day after school, Arjun would sit with his books and explore different topics. He was especially fascinated by science and technology. He wanted to understand how machines think and whether they could become as intelligent as humans.\n\nOne day, while browsing through an old computer at school, Arjun discovered the concept of artificial intelligence. He read about how machines can learn from data, recognize patterns, and even make decisions. This idea completely amazed him.\n\nHe started imagining a future where machines could help humans solve complex problems. That day, he made a decision that he wanted to become an AI engineer.\n\nArjun began learning programming using free resources. At first, everything seemed difficu

'Once upon a time, there was a young boy named Arjun who lived in a small town surrounded by hills and rivers. He was very curious about how things worked and always asked questions. His parents often smiled at his curiosity and encouraged him to learn more.\n\nEvery day after school, Arjun would sit with his books and explore different topics. He was especially fascinated by science and technology. He wanted to understand how machines think and whether they could become as intelligent as humans.\n\nOne day, while browsing through an old computer at school, Arjun discovered the concept of artificial intelligence. He read about how machines can learn from data, recognize patterns, and even make decisions. This idea completely amazed him.\n\nHe started imagining a future where machines could help humans solve complex problems. That day, he made a decision that he wanted to become an AI engineer.\n\nArjun began learning programming using free resources. At first, everything seemed difficult. He struggled to understand basic concepts and often felt frustrated. But his curiosity kept him going.\n\nSlowly, he started building small programs. He created simple calculators, then moved on to small games. Each small success gave him confidence. He realized that learning takes time, but consistency is the key.\n\nHis teachers noticed his interest and guided him whenever he was stuck. Arjun also started reading about data, algorithms, and machine learning. He learned that data is the most important part of building intelligent systems.\n\nEven though he faced many challenges, he never gave up. He kept learning, practicing, and improving every single day.'

In [31]:
sent_tokenize(corpus)

['Once upon a time, there was a young boy named Arjun who lived in a small town surrounded by hills and rivers.',
 'He was very curious about how things worked and always asked questions.',
 'His parents often smiled at his curiosity and encouraged him to learn more.',
 'Every day after school, Arjun would sit with his books and explore different topics.',
 'He was especially fascinated by science and technology.',
 'He wanted to understand how machines think and whether they could become as intelligent as humans.',
 'One day, while browsing through an old computer at school, Arjun discovered the concept of artificial intelligence.',
 'He read about how machines can learn from data, recognize patterns, and even make decisions.',
 'This idea completely amazed him.',
 'He started imagining a future where machines could help humans solve complex problems.',
 'That day, he made a decision that he wanted to become an AI engineer.',
 'Arjun began learning programming using free resources.',


In [32]:
raw_sent = sent_tokenize(corpus)

In [33]:
for i in raw_sent:
    print(simple_preprocess(i))

['once', 'upon', 'time', 'there', 'was', 'young', 'boy', 'named', 'arjun', 'who', 'lived', 'in', 'small', 'town', 'surrounded', 'by', 'hills', 'and', 'rivers']
['he', 'was', 'very', 'curious', 'about', 'how', 'things', 'worked', 'and', 'always', 'asked', 'questions']
['his', 'parents', 'often', 'smiled', 'at', 'his', 'curiosity', 'and', 'encouraged', 'him', 'to', 'learn', 'more']
['every', 'day', 'after', 'school', 'arjun', 'would', 'sit', 'with', 'his', 'books', 'and', 'explore', 'different', 'topics']
['he', 'was', 'especially', 'fascinated', 'by', 'science', 'and', 'technology']
['he', 'wanted', 'to', 'understand', 'how', 'machines', 'think', 'and', 'whether', 'they', 'could', 'become', 'as', 'intelligent', 'as', 'humans']
['one', 'day', 'while', 'browsing', 'through', 'an', 'old', 'computer', 'at', 'school', 'arjun', 'discovered', 'the', 'concept', 'of', 'artificial', 'intelligence']
['he', 'read', 'about', 'how', 'machines', 'can', 'learn', 'from', 'data', 'recognize', 'patterns',

In [34]:
#doing same for both story with a logic

folder_path = r"E:\Gen AI Bootcamp\Full-Stack-GenerativeAI-And-Agentic-AI-Bootcamp\Class 06_11April26 word2vec, avg word2vec practical\Data"
story =[]
for filename in os.listdir(folder_path):
  file_path=os.path.join(folder_path,filename)
  with open(file_path,encoding='unicode_escape') as f:
    corpus=f.read()
  raw_sent=sent_tokenize(corpus)
  for sent in raw_sent:
    story.append(simple_preprocess(sent))

In [35]:
story

[['once',
  'upon',
  'time',
  'there',
  'was',
  'young',
  'boy',
  'named',
  'arjun',
  'who',
  'lived',
  'in',
  'small',
  'town',
  'surrounded',
  'by',
  'hills',
  'and',
  'rivers'],
 ['he',
  'was',
  'very',
  'curious',
  'about',
  'how',
  'things',
  'worked',
  'and',
  'always',
  'asked',
  'questions'],
 ['his',
  'parents',
  'often',
  'smiled',
  'at',
  'his',
  'curiosity',
  'and',
  'encouraged',
  'him',
  'to',
  'learn',
  'more'],
 ['every',
  'day',
  'after',
  'school',
  'arjun',
  'would',
  'sit',
  'with',
  'his',
  'books',
  'and',
  'explore',
  'different',
  'topics'],
 ['he',
  'was',
  'especially',
  'fascinated',
  'by',
  'science',
  'and',
  'technology'],
 ['he',
  'wanted',
  'to',
  'understand',
  'how',
  'machines',
  'think',
  'and',
  'whether',
  'they',
  'could',
  'become',
  'as',
  'intelligent',
  'as',
  'humans'],
 ['one',
  'day',
  'while',
  'browsing',
  'through',
  'an',
  'old',
  'computer',
  'at',
  'sc

In [36]:
story[0]

['once',
 'upon',
 'time',
 'there',
 'was',
 'young',
 'boy',
 'named',
 'arjun',
 'who',
 'lived',
 'in',
 'small',
 'town',
 'surrounded',
 'by',
 'hills',
 'and',
 'rivers']

In [37]:
story[2]

['his',
 'parents',
 'often',
 'smiled',
 'at',
 'his',
 'curiosity',
 'and',
 'encouraged',
 'him',
 'to',
 'learn',
 'more']

In [38]:
#by default#

# Word2Vec(..., sg=0)  # CBOW #by default
# Word2Vec(..., sg=1)  # Skip-gram

In [ ]:


#custom_model = gensim.models.Word2Vec(
#    min_count= 5,   # min word count data remove
#    vector_size= 150, 
#)

## These all the values are the Hyperparameter (we can change it)
## sg not mension so, by default it CBOW

In [55]:


custom_model = gensim.models.Word2Vec(
    window= 10,
    min_count= 1,
    vector_size= 150, 
)

## These all the values are the Hyperparameter (we can change it)
## sg not mension so, by default it CBOW

In [57]:
custom_model.build_vocab(story)

In [ ]:
custom_model.corpus_count

44

In [62]:
custom_model.train(story, total_examples=custom_model.corpus_count, epochs = 10)  ## epoche is hyperpyarameter, it could be anything

(3409, 5220)

(3409, 5220)

part   Meaning

3409 = number of documents

5220 = number of unique words

In [63]:
## wv = word vector

custom_model.wv["arjun"]

array([-0.00015591, -0.0010703 ,  0.0060785 , -0.00600821, -0.00242359,
       -0.00430484, -0.00932003,  0.01058115, -0.00965235,  0.0030909 ,
        0.00798179, -0.00397186, -0.00753484,  0.00075761, -0.01014617,
       -0.00806525,  0.00795864,  0.00672468, -0.00923542,  0.01372813,
       -0.01093896,  0.00394161,  0.01161107,  0.01086326,  0.00315523,
       -0.00106717, -0.01125761, -0.00176534, -0.0009803 , -0.01672976,
       -0.00193171, -0.00598437, -0.00549292, -0.01037003, -0.01285857,
       -0.00211017,  0.02279315,  0.00520776, -0.00637331, -0.01275247,
        0.00174563, -0.00415141, -0.00777965, -0.00581964,  0.00579979,
       -0.00639253,  0.01026587,  0.00043498,  0.00297434,  0.00225605,
       -0.00647229,  0.00112426, -0.00281992, -0.0003273 , -0.0075786 ,
        0.00291675,  0.0045968 ,  0.00821226, -0.00568095, -0.00081421,
       -0.01056663, -0.01599464, -0.00841113, -0.01000035,  0.00312521,
       -0.00846909, -0.00807008, -0.01546639, -0.01858189, -0.01

In [64]:
len(custom_model.wv["arjun"])  # same as vector size which we use while creating a custom model


150

In [67]:
custom_model.wv["ai"]

array([ 5.22753038e-03,  4.48844396e-03, -4.37989132e-03, -2.15482403e-04,
       -6.44205278e-03, -4.58936393e-03, -4.15878743e-03,  1.24506494e-02,
       -6.64509740e-03,  6.43713493e-03, -1.28220301e-03, -3.99726676e-03,
        2.03186041e-03,  2.87840585e-03, -4.91173193e-03,  2.40003341e-03,
        4.30653291e-03, -1.25815184e-03,  1.81022659e-03, -3.05350474e-03,
       -8.02469673e-04, -6.00303197e-03,  5.14317956e-03, -3.08049330e-03,
       -1.17983657e-03,  4.40383516e-03, -1.53423299e-03, -2.34048255e-03,
       -6.67976821e-03, -1.01045575e-02,  5.71746798e-03, -6.49737287e-03,
       -1.41728192e-03, -1.33816618e-03, -3.99016729e-03,  5.06368838e-03,
        8.47703498e-03,  6.45090779e-03,  3.47314868e-03, -8.80981609e-03,
       -2.61157751e-03, -6.13562239e-04, -6.67524477e-03, -4.31443518e-03,
        5.63913304e-03,  4.97210585e-03,  1.40058619e-04,  4.20478964e-03,
       -5.78934187e-03, -4.43867315e-03, -3.02627264e-03,  6.36008335e-03,
        3.66249005e-04,  

In [71]:
custom_model.wv.most_similar("arjun")

[('and', 0.8689543604850769),
 ('to', 0.8469886183738708),
 ('his', 0.8370557427406311),
 ('that', 0.835411787033081),
 ('he', 0.8230711817741394),
 ('of', 0.8081210851669312),
 ('in', 0.8042126893997192),
 ('was', 0.7866895198822021),
 ('could', 0.7570267915725708),
 ('simple', 0.7471853494644165)]

In [72]:
custom_model.wv.most_similar("ai")


[('and', 0.5983321070671082),
 ('the', 0.5812115669250488),
 ('computer', 0.5687210559844971),
 ('to', 0.5618873238563538),
 ('weather', 0.5580054521560669),
 ('he', 0.5509629249572754),
 ('arjun', 0.5488499999046326),
 ('that', 0.5413717031478882),
 ('where', 0.5382276773452759),
 ('his', 0.527862548828125)]

In [73]:
custom_model.wv.most_similar("engineer")



[('in', 0.49149614572525024),
 ('and', 0.46347999572753906),
 ('was', 0.45773187279701233),
 ('of', 0.4509519636631012),
 ('his', 0.44717901945114136),
 ('he', 0.4383416771888733),
 ('young', 0.43643149733543396),
 ('hard', 0.4360801875591278),
 ('to', 0.4348940849304199),
 ('curiosity', 0.43317607045173645)]

In [74]:
print("arjun" in custom_model.wv)
print("ai" in custom_model.wv)
print("engineer" in custom_model.wv)
print(custom_model.wv.index_to_key[:50])

True
True
True
['and', 'he', 'his', 'arjun', 'to', 'that', 'the', 'was', 'in', 'started', 'him', 'small', 'data', 'of', 'learning', 'day', 'more', 'by', 'could', 'time', 'but', 'about', 'work', 'curiosity', 'better', 'as', 'at', 'for', 'not', 'farmers', 'this', 'world', 'real', 'their', 'system', 'help', 'many', 'using', 'machines', 'weather', 'kept', 'model', 'problems', 'very', 'make', 'on', 'how', 'humans', 'gave', 'intelligent']


In [76]:
custom_model.wv.doesnt_match(["arjun", "engineer", "programming", "farmers"])

'engineer'

In [ ]:
# print(f"Stories loaded: {len(story)}")
# print(f"Sample: {story[:2]}")

Stories loaded: 44
Sample: [['once', 'upon', 'time', 'there', 'was', 'young', 'boy', 'named', 'arjun', 'who', 'lived', 'in', 'small', 'town', 'surrounded', 'by', 'hills', 'and', 'rivers'], ['he', 'was', 'very', 'curious', 'about', 'how', 'things', 'worked', 'and', 'always', 'asked', 'questions']]


In [ ]:
# for i in range(5):
#     model = gensim.models.Word2Vec(
#         story, window=10, min_count=1, vector_size=150, epochs=10
#     )
#     print(f"Run {i+1}:", model.wv.most_similar("engineer"))

Run 1: [('programs', 0.2385317087173462), ('lose', 0.2195308655500412), ('algorithms', 0.21495571732521057), ('practicing', 0.20005598664283752), ('who', 0.1644534021615982), ('from', 0.1642623394727707), ('in', 0.1638709157705307), ('curiosity', 0.1629614681005478), ('through', 0.15496230125427246), ('recognize', 0.15426163375377655)]
Run 2: [('programs', 0.2385317087173462), ('lose', 0.2195308655500412), ('algorithms', 0.21495571732521057), ('practicing', 0.20005598664283752), ('who', 0.1644534021615982), ('from', 0.1642623394727707), ('in', 0.1638709157705307), ('curiosity', 0.1629614681005478), ('through', 0.15496230125427246), ('recognize', 0.15426163375377655)]
Run 3: [('programs', 0.2385317087173462), ('lose', 0.2195308655500412), ('algorithms', 0.21495571732521057), ('practicing', 0.20005598664283752), ('who', 0.1644534021615982), ('from', 0.1642623394727707), ('in', 0.1638709157705307), ('curiosity', 0.1629614681005478), ('through', 0.15496230125427246), ('recognize', 0.154261